operator/identitygc/gc.go

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package identitygc
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cilium/workerpool"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/allocator"
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	ciliumV2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/rate"
+)
+
+// params contains all the dependencies for the identity-gc.
+// They will be provided through dependency injection.
+type params struct {
+	cell.In
+
+	Logger    logrus.FieldLogger
+	Lifecycle hive.Lifecycle
+
+	Clientset k8sClient.Clientset
+	Identity  resource.Resource[*v2.CiliumIdentity]
+
+	Cfg       Config
+	SharedCfg SharedConfig
+}
+
+// GC represents the Cilium identities periodic GC.
+type GC struct {
+	logger logrus.FieldLogger
+
+	clientset ciliumV2.CiliumIdentityInterface
+	identity  resource.Resource[*v2.CiliumIdentity]
+
+	allocationMode string
+
+	gcInterval       time.Duration
+	heartbeatTimeout time.Duration
+	gcRateInterval   time.Duration
+	gcRateLimit      int64
+
+	wp             *workerpool.WorkerPool
+	heartbeatStore *heartbeatStore
+	mgr            *controller.Manager
+
+	// rateLimiter is meant to rate limit the number of
+	// identities being GCed by the operator. See the documentation of
+	// rate.Limiter to understand its difference than 'x/time/rate.Limiter'.
+	//
+	// With our rate.Limiter implementation Cilium will be able to handle bursts
+	// of identities being garbage collected with the help of the functionality
+	// provided by the 'policy-trigger-interval' in the cilium-agent. With the
+	// policy-trigger even if we receive N identity changes over the interval
+	// set, Cilium will only need to process all of them at once instead of
+	// processing each one individually.
+	rateLimiter *rate.Limiter
+
+	allocationCfg identityAllocationConfig
+	allocator     *allocator.Allocator
+
+	enableMetrics bool
+	// counters for GC failed/successful runs
+	failedRuns     float64
```

Comment:
these should probably just be ints

Reply:
I've used floats because that's what `(prometheus.Gauge).Set` requires, but you're right, this may feel confusing. Better to change this and cast that at `Set` call site.